homeassistant/components/dlna_dmr/media_player.py

```diff
@@ -586,10 +593,30 @@ async def async_play_media(
         """Play a piece of media."""
         _LOGGER.debug("Playing media: %s, %s, %s", media_type, media_id, kwargs)
         assert self._device is not None
+
+        didl_metadata: str | None = None
+        title: str = ""
+
+        # If media is media_source, resolve it to url and MIME type, and maybe metadata
+        if media_source.is_media_source_id(media_id):
+            sourced_media = await media_source.async_resolve_media(self.hass, media_id)
+            media_type = sourced_media.mime_type
+            media_id = sourced_media.url
+            _LOGGER.debug("sourced_media is %s", sourced_media)
+            if sourced_metadata := getattr(sourced_media, "didl_metadata", None):
```

Comment:
Can you use `isinstance` to detect if you're browsing UPNP media?

Reply:
I could (and did originally), but that then needs an explicit import of the `dlna_dms` component, and therefore a dependency within the dlna_dmr manifest. This current approach is also a bit more flexible because it can handle other integrations extending `PlayMedia` with a `didl_metadata` attribute, and they don't have to depend on `dlna_dms` for it to work.

This has also led me to think about how media sources can provide metadata to media players more generally. I'll work up an architecture discussion for that next week.